File: hard/768.max-chunks-to-make-sorted-ii.java
/*
 * @lc app=leetcode id=768 lang=java
 *
 * [768] Max Chunks To Make Sorted II
 */

// @lc code=start
class Solution {
    public int maxChunksToSorted(int[] arr) {
        int len = arr.length;
        if (len == 1) {
            return 1;
        }

        int[] maxFromLeft = new int[len];
        int[] minFromRight = new int[len];

        maxFromLeft[0] = arr[0];
        for (int i=1; i<len; i++) {
            maxFromLeft[i] = Math.max(maxFromLeft[i-1], arr[i]);
        }

        minFromRight[len-1] = arr[len-1];
        for (int i=len-2; i>=0; i--) {
            minFromRight[i] = Math.min(minFromRight[i+1], arr[i]);
        }

        int count = 1;
        for (int i=0; i<len-1; i++) {
            if (maxFromLeft[i] <= minFromRight[i+1]) {
                count++;
            }
        }

        // [0,1]
        // [1,0,2,3,4]
        return count;
    }
}
// @lc code=end

File: hard/768.max-chunks-to-make-sorted-ii.js
/*
 * @lc app=leetcode id=768 lang=javascript
 *
 * [768] Max Chunks To Make Sorted II
 */

// @lc code=start
/**
 * @param {number[]} arr
 * @return {number}
 */
var maxChunksToSorted = function(arr) {
  let monoStack = [];
  for (let num of arr) {
    if (monoStack.length > 0 && num < monoStack[monoStack.length - 1]) {
      let top = monoStack[monoStack.length - 1];
      while (monoStack.length > 0 && num < monoStack[monoStack.length - 1]) {
        monoStack.pop();
      }
      monoStack.push(top);
    } else {
      monoStack.push(num);
    }
  }

  console.log(monoStack);

  //[1,1,0,0,1]
  // [0,2,1,3,4,4]
  return monoStack.length;
};
// @lc code=end

